src/common/KokkosKernels_SparseUtils.hpp

```diff
@@ -1955,6 +1963,167 @@ void kk_reduce_numrows_larger_than_threshold(
+
+  KOKKOS_INLINE_FUNCTION
+  size_type row_offset() { return row_off; }
+
+ protected:
+  lno_t block_size;  // = num_block_cols = num_block_rows
+  lno_t row_begin;
+  lno_t row_end;
+
+ private:  // cache
+  size_type row_off;
+  lno_t row_size;
+};
+
+template <SparseMatrixFormat /* format */, typename lno_t, typename size_type>
+class MatrixRowIndex;
+
+template <typename lno_t, typename size_type>
+class MatrixRowIndex<BlockCRS, lno_t, size_type>
+    : public RowIndexBase<lno_t, size_type> {
+ public:
+  using Base = RowIndexBase<lno_t, size_type>;
+
+  KOKKOS_INLINE_FUNCTION
+  MatrixRowIndex(const lno_t block_size_, const lno_t row_begin_,
+                 const lno_t row_end_)
+      : Base(block_size_, row_begin_, row_end_) {}
+
+  KOKKOS_INLINE_FUNCTION
+  size_type block(const lno_t col_idx) {
+    return Base::row_offset() + col_idx * Base::block_size;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  size_type block_stride() { return Base::size() * Base::block_size; }
+
+  KOKKOS_INLINE_FUNCTION
+  size_type value(const lno_t col_idx, const lno_t block_row,
+                  const lno_t block_col) {
+    return block(col_idx) + block_row * block_stride() + block_col;
+  }
+};
+
+template <typename lno_t, typename size_type>
+class MatrixRowIndex<BSR, lno_t, size_type>
+    : public RowIndexBase<lno_t, size_type> {
+ public:
+  using Base = RowIndexBase<lno_t, size_type>;
+
+  KOKKOS_INLINE_FUNCTION
+  MatrixRowIndex(const lno_t block_size_, const lno_t row_begin_,
+                 const lno_t row_end_)
+      : Base(block_size_, row_begin_, row_end_) {}
+
+  KOKKOS_INLINE_FUNCTION
+  size_type block(const lno_t col_idx) {
+    return Base::row_offset() + col_idx * Base::block_mtx_size();
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  size_type block_stride() { return Base::block_size; }
+
+  KOKKOS_INLINE_FUNCTION
+  size_type value(const lno_t col_idx, const lno_t block_row,
+                  const lno_t block_col) {
+    return block(col_idx) + block_row * block_stride() + block_col;
+  }
+};
+
+template <typename mtx_t>
+struct MatrixTraits;
+
+template <typename scalar_t, typename lno_t, typename device,
+          typename mem_traits, typename size_type>
+struct MatrixTraits<
+    KokkosSparse::CrsMatrix<scalar_t, lno_t, device, mem_traits, size_type>> {
+  static constexpr auto format = KokkosKernels::BlockCRS;
+};
+
+template <typename scalar_t, typename lno_t, typename device,
+          typename mem_traits, typename size_type>
+struct MatrixTraits<KokkosSparse::Experimental::BsrMatrix<
+    scalar_t, lno_t, device, mem_traits, size_type>> {
+  static constexpr auto format = KokkosKernels::BSR;
+};
+
+template <SparseMatrixFormat /* outFormat */>
+struct MatrixConverter;
+
+template <>
+struct MatrixConverter<BlockCRS> {
+  template <
+      typename scalar_t, typename lno_t, typename device, typename size_type,
+      typename crsMat_t =
+          KokkosSparse::CrsMatrix<scalar_t, lno_t, device, void, size_type>>
+  static crsMat_t from_crs(
```

Comment:
Why does ``MatrixConverter<BlockCRS>::from_crs`` return another CrsMatrix (not BlockCrs), but ``MatrixConverter<BSR>::from_crs`` returns a BsrMatrix?

Reply:
Because the test operates on block-formatted `CrsMatrix`, not `BlockCrsMatrix` and I decided to leave matrix generating utilities intact... Looking back at this, I'd agree that it's an inconsistent design for a generic converter. Shall we update the test to operate on `BlockCrsMatrix` and make it the return type for the converter ?

Reply:
If that's not too much work, it would be nice to make this consistent.

Reply:
Fixed on 14927f32ec32aca4a0b16585eba0902613109a3c
Was much easier that it seemed - thank you for pointing this out.